daml-assistant/src/DA/Daml/Assistant/Install.hs

```diff
@@ -300,20 +307,34 @@ extractAndInstall env source =
     where throwError msg e = liftIO $ throwIO $ assistantErrorBecause ("Invalid SDK release: " <> msg) e
 
 -- | Download an sdk tarball and install it.
-httpInstall :: InstallEnv -> InstallURL -> IO ()
-httpInstall env@InstallEnv{..} (InstallURL url) = do
+httpInstall :: InstallEnv -> SdkVersion -> IO ()
+httpInstall env@InstallEnv{..} version = do
     unlessQuiet env $ output "Downloading SDK release."
-    request <- requiredAny "Failed to parse HTTPS request." $ parseRequest ("GET " <> unpack url)
-    requiredAny "Failed to download SDK release." $ withResponse request $ \response -> do
-        when (getResponseStatusCode response /= 200) $
-            throwIO . assistantErrorBecause "Failed to download release."
-                    . pack . show $ getResponseStatus response
-        let totalSizeM = readMay . BS.UTF8.toString =<< headMay
-                (getResponseHeader "Content-Length" response)
-        extractAndInstall env
-            . maybe id (\s -> (.| observeProgress s)) totalSizeM
-            $ getResponseBody response
+    requiredAny "Failed to download SDK release." $
+        firstNon404 (map tryLocation (maybeToList (fmap (Artifactory.versionLocation version) artifactoryApiKeyM) <> [Github.versionLocation version]))
```

Comment:
IIUC, when the EE SDK release isn't available on artifactory, it installs the CE SDK from github instead.

Is this really desirable behavior? Would it be better to just fail, and let the user know that the EE release couldn't be found, and give them a command-line that lets them install the CE instead?

I'm (also) concerned people will get stuck with the wrong version by accident.

Reply:
Older versions (before the most recent snapshot) don’t have a separate EE edition. I think asking users to be aware since when we started publishing the enterprise edition is too annoying.

That leaves us with two options. Either we do what we do here and try artifactory first and go to GH releases on 404s. The alternative would be to hardcode the version number of the first snapshot and only try artifactory for newer versions and only try GH releases for older versions.

I don’t feel too strongly about this. Hardcoding the version in the assistant seemed a bit uglier and less robust but if you disagree, I’m happy to switch to that model.

Reply:
I prefer hardcoding the version number for the first release that's available on artifactory. Or alternatively, have a cut-off point at which it will no longer use Github as an automatic backup. I think these will lead to better behaviour for our users.

Reply:
changed in 435c5b2208e1191e3724d02fa2794aae50644b6b

Reply:
Thanks!